pkg/cloud/services/ec2/instances.go

```diff
@@ -321,43 +321,43 @@ func (s *Service) findSubnet(scope *scope.MachineScope) (string, error) {
 	case failureDomain != nil:
 		if scope.AWSMachine.Spec.PublicIP != nil && *scope.AWSMachine.Spec.PublicIP {
 			subnets := s.scope.Subnets().FilterPublic().FilterByZone(*failureDomain)
-			if len(subnets) == 0 {
-				errMessage := fmt.Sprintf("failed to run machine %q with public IP, no public subnets available in availability zone %q",
-					scope.Name(), *failureDomain)
-				record.Warnf(scope.AWSMachine, "FailedCreate", errMessage)
-				return "", awserrors.NewFailedDependency(errMessage)
+			if subnetID := getSubnetID(subnets); subnetID != "" {
```

Comment:
Instead of returning early when the ID is found, checking negative case first and returning early if fails is a more readable pattern to use and by this way calling getSubnetId() twice in the same case statement will not be needed.